### ISSUE TITLE
Support for Knative Service

### ISSUE THREAD

**Issue by piomin** [skupperproject/skupper#1192]
Currently, Skupper is able to expose deployment, pods, statefulset, service, or OpenShift deploymentconfig. It would be nice to support Knative Service the same as those objects.

**Comment by grs** [comment]
With the cli, you can do `skupper expose service xxxx` for a knative service. (Skupper doesn't read any annotations from knative service resources, but the plan is to move away from annotations anyway, with a more explicit definition of the connectors and listeners).

**Comment by piomin** [comment]
Yes, I can do it. But to clarify. I have the following Knative Service:
```
$ oc get ksvc
NAME                   URL                                                          LATESTCREATED                LATESTREADY                  READY   REASON
sample-kotlin-spring   http://sample-kotlin-spring.interconnect.svc.cluster.local   sample-kotlin-spring-00001   sample-kotlin-spring-00001   True
```

It corresponds to the following list of Kubernetes Services:
```
$ oc get svc
sample-kotlin-spring                 ExternalName   <none>           kourier-internal.knative-serving-ingress.svc.cluster.local   80/TCP                                               25h
sample-kotlin-spring-00001           ClusterIP      172.30.85.49     <none>                                                       80/TCP,443/TCP                                       25h
sample-kotlin-spring-00001-private   ClusterIP      172.30.254.180   <none>                                                       80/TCP,443/TCP,9090/TCP,9091/TCP,8022/TCP,8012/TCP   25h
```

When I'm calling the skupper to expose service I'm not exposing Knative Service, but the Kubernetes Service (`type=ExternalName`). However, even if I do that as shown below:
```
$ skupper expose service sample-kotlin-spring --address sample-kotlin-spring
service sample-kotlin-spring exposed as sample-kotlin-spring
```

Now, the status of my Knative Service is:
```
$ oc get ksvc
NAME                   URL                                                          LATESTCREATED                LATESTREADY                  READY   REASON
sample-kotlin-spring   http://sample-kotlin-spring.interconnect.svc.cluster.local   sample-kotlin-spring-00001   sample-kotlin-spring-00001   False   NotOwned
```

Detailed reason:
_There is an existing placeholder Endpoint "sample-kotlin-spring" that we do not own._

I also tried e.g. to create a service with Skupper:
```
$ skupper service create <name> <port>
```

And then bind it to the `sample-kotlin-spring` with the command
```
$ skupper service bind <service-name> <target-type> <target-name>
```

It doesn't modify the service managed by Knative, and therefore Knative Service still works. But this time when I'm trying to call the service through the Service created by Skupper it is not able to connect it (I goes that because of the `ExternalName` type)

**Comment by grs** [comment]
If you do `skupper expose service sample-kotlin-spring --address sample-kotlin-spring-skupper` (or some other separate name for the skupper created service), it should work. What do you see in the skupper-services configmap in that case?

**Comment by piomin** [comment]
Well, that's something similar to the last two commands I placed in my previous message. But ok - I'm doing smth like that on my first cluster:
```
$ skupper expose service sample-kotlin-spring --address sample-kotlin-spring-skupper
```

It creates a Skupper service and exposes it to the linked cluster. I'm switching to that cluster and exposing it as a `Route` (alternatively I was trying to port-forward service directly - but the result is still the same)
```
$ oc expose svc/sample-kotlin-spring-skupper
```

Then I'm calling a `Route`:
```
$ curl http://sample-kotlin-spring-skupper-interconnect.apps.cluster-44jwm.44jwm.sandbox1320.opentlc.com/persons
```

The result is HTTP 404 not found. 

Skupper service and network status:
```
$ skupper service status
Services exposed through Skupper:
╰─ sample-kotlin-spring-skupper (tcp port 80)

$ skupper network status
Sites:
├─ [local] b5212fc - interconnect
│  URL: skupper-inter-router-interconnect.apps.pvxvtsz4.eastus.aroapp.io
│  mode: interior
│  name: interconnect
│  namespace: interconnect
│  version: 1.4.2
│  ╰─ Services:
│     ╰─ name: sample-kotlin-spring-skupper
│        address: sample-kotlin-spring-skupper: 80
│        protocol: tcp
╰─ [remote] 903f498 - interconnect
   name: interconnect
   namespace: interconnect
   sites linked to: b5212fc-interconnect
   version: 1.4.2
   ╰─ Services:
      ╰─ name: sample-kotlin-spring-skupper
         address: sample-kotlin-spring-skupper: 80
         protocol: tcp
         ╰─ Targets:
            ╰─ name: sample-kotlin-spring.interconnect
```

**Comment by grs** [comment]
I suspect that the issue there is the Host header is not matching anything when the request is processed by the knative server. You could verify that by setting the Host header to 'sample-kotlin-spring' in the curl request.

To be able to keep the same name, try `skupper expose service sample-kotlin-spring --address sample-kotlin-spring --enable-ingress-from-target-site Never` That will prevent it overwriting the service in the namespace in which the service already exists, but will use the same service name in the remote site.

**Comment by piomin** [comment]
Yes, with the `--enable-ingress-from-target-site` parameter I'm able to keep the same name and it won't break the Knative Service. Skupper creates a Service on the linked cluster. Then I'm enabling port-forward for that service:
```
$ oc port-forward svc/sample-kotlin-spring 8080:80
```

Adding the `Host` header won't change anything. It still receives HTTP 404. 
```
$ curl http://localhost:8080/persons -H "Host: sample-kotlin-spring" -v
*   Trying 127.0.0.1:8080...
* Connected to localhost (127.0.0.1) port 8080 (#0)
> GET /persons HTTP/1.1
> Host: sample-kotlin-spring
> User-Agent: curl/7.88.1
> Accept: */*
>
< HTTP/1.1 404 Not Found
< date: Thu, 03 Aug 2023 15:05:00 GMT
< server: envoy
< content-length: 0
<
* Connection #0 to host localhost left intact
```

Exactly the same app deployed as a standard `Deployment` works perfectly fine with Skupper.

**Comment by grs** [comment]
Does it work if you use a fully qualified service name (or at least qualified by namespace) as the Host header?

**Comment by piomin** [comment]
Yes, it works! :) With `port-forwarding` I can set that header. But in fact, it is a workaround, right? For example, if I have that service exposed with a `Route`, I cannot change the `Host` header when calling it.  

```
$ curl http://localhost:8080/persons -H "Host: sample-kotlin-spring.interconnect" -v
*   Trying 127.0.0.1:8080...
* Connected to localhost (127.0.0.1) port 8080 (#0)
> GET /persons HTTP/1.1
> Host: sample-kotlin-spring.interconnect
> User-Agent: curl/7.88.1
> Accept: */*
>
< HTTP/1.1 200 OK
< content-type: application/json
< date: Thu, 03 Aug 2023 21:46:15 GMT
< x-envoy-upstream-service-time: 28741
< server: envoy
< transfer-encoding: chunked
<
* Connection #0 to host localhost left intact
[{"id":1,"name":"John Smith","age":39,"gender":"MALE"},{"id":2,"name":"Paul Walker","age":48,"gender":"MALE"},{"id":3,"name":"Kate Morgan","age":21,"gender":"FEMALE"}]
```

Could you also take a look at that issue: https://github.com/skupperproject/skupper/issues/1194? Assuming, I'm able to use Skupper with Knative Service I would like to lb between the services across e.g. two clusters with auto-scaling enabled.

**Comment by grs** [comment]
The only way skupper can re-write the Host header is if you use `--protocol http` when you do `skupper expose`. You should also then use the qualified service name as the target. This would also give request level load balancing.

**Comment by piomin** [comment]
Request level load balancing - you mean round-robin lb between all the pods?

**Comment by grs** [comment]
The request load balancing seeks to minimise latency. So if you have pods on one site that are able to handle more load than a single pod on another site, that first site should get more requests. If one site is slower, it will also get fewer requests.

**Comment by piomin** [comment]
Don't plan at least enable (not by default) round-robin lb?

**Comment by grs** [comment]
We certainly plan to support more control over loadbalancing with --protocol tcp, but that is not a trivial change and so will take some time to materialise.

Can you describe your requirement for round robin in more detail?

**Comment by piomin** [comment]
It can also be http. Load balancing is based on latency. It's cool, but the problem for me is - if I have two clusters, with a different number of app pods, e.g. there is 1 pod in cluster `c1`, and three pods in the `c2` cluster (latency is similar in all four pods) -> load balancing is 50%-50% between the clusters, what means that pod on cluster `c1` receives 50%, where a single pod in `c2` receives ~17% of total traffic. Is it possible to have ~25% per pod in this scenario? I guess it can be problematic for Skupper to do it since Skupper lb is based on Kubernetes Service?

**Comment by piomin** [comment]
@grs any thoughts about it?
